nexus-repository-r-it/src/test/java/org/sonatype/nexus/plugins/r/internal/RHostedIT.java

```diff
@@ -0,0 +1,131 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.ops4j.pax.exam.Configuration;
+import org.ops4j.pax.exam.Option;
+
+import static org.apache.commons.compress.compressors.CompressorStreamFactory.GZIP;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isEmptyString;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+public class RHostedIT
+    extends RITSupport
+{
+  private RClient client;
+
+  private Repository repository;
+
+  @Rule
+  public RepositoryRuleR repos = new RepositoryRuleR(() -> repositoryManager);
+
+  @Configuration
+  public static Option[] configureNexus() {
+    return NexusPaxExamSupport.options(
+        NexusITSupport.configureNexusBase(),
+        nexusFeature("org.sonatype.nexus.plugins", "nexus-repository-r")
+    );
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    BaseUrlHolder.set(this.nexusUrl.toString());
+    repository = repos.createRHosted("r-hosted-test");
+    client = createRHostedClient(repository);
+  }
+
+  @Test
+  public void testPackagesProcessing() throws Exception
+  {
+    assertThat(getAllComponents(repository), hasSize(0));
+    final File file = testData.resolveFile(AGRICOLAE_PKG_FILE_NAME_131_TGZ);
+    client.put(AGRICOLAE_PATH_FULL_131_TGZ, new ByteArrayEntity(Files.readAllBytes(Paths.get(file.getAbsolutePath()))));
+
+    //Verify DB contains data about uploaded component and asset
+    Component component = findComponent(repository, AGRICOLAE_PKG_NAME);
+    assertThat(component.name(), is(equalTo(AGRICOLAE_PKG_NAME)));
+    assertThat(component.version(), is(equalTo(AGRICOLAE_PKG_VERSION_131)));
+
+    //Verify Asset is created.
+    Asset asset = findAsset(repository, AGRICOLAE_PATH_FULL_131_TGZ);
+    assertThat(asset.name(), is(equalTo(AGRICOLAE_PATH_FULL_131_TGZ)));
+    assertThat(asset.format(), is(equalTo(R_FORMAT_NAME)));
+
+    //Verify that package could be downloaded from NXRM
+    HttpResponse resp = client.fetch(AGRICOLAE_PATH_FULL_131_TGZ);
+    assertThat(resp.getEntity().getContentType().getValue(), equalTo(CONTENT_TYPE_TGZ));
+    assertSuccessResponseMatches(client.fetch(AGRICOLAE_PATH_FULL_131_TGZ), AGRICOLAE_PKG_FILE_NAME_131_TGZ);
+
+    //TODO DELETE ASSET and check that component is deleted. will be implemented after NEXUS-20711 fix.
+  }
+
+  @Test
+  public void testMetadataProcessing() throws Exception
+  {
+    final File expectedPackageFile = testData.resolveFile(PACKAGES_AGRICOLAE_131_NAME);
+    final String expectedPackageData = new String(Files.readAllBytes(expectedPackageFile.toPath()));
+
+    assertThat(getAllComponents(repository), hasSize(0));
+
+    final File file = testData.resolveFile(AGRICOLAE_PKG_FILE_NAME_131_TGZ);
+    client.put(AGRICOLAE_PATH_FULL_131_TGZ, new ByteArrayEntity(Files.readAllBytes(Paths.get(file.getAbsolutePath()))));
```

Comment:
Suggestion: This could also re-use the put package method I was suggesting above

Reply:
ec3e07bd99b71ebdd951f01522ad4fd1a3ff0373